Add DNS Zone support

raymens opened the pull request:
This PR closes issue #270

The changes in this PR are as follows:

* Add DNS Zone support.

I have read the [contributing guidelines](CONTRIBUTING.md) and have completed the following:

* [x] **Tested my code** end-to-end against a live Azure subscription.
* [x] **Updated the documentation** in the docs folder for the affected changes.
* [x] **Written unit tests** against the modified code that I have made.
* [x] **Updated the [release notes](RELEASE_NOTES.md)** with a new entry for this PR.
* [x] **Checked the coding standards** outlined in the contributions guide and ensured my code adheres to them.

If I haven't completed any of the tasks above, I include the reasons why here:

raymens commented:
```fsharp
let dns = dnsZone {
    name "server"
    add_records [
        dnsRecord {
            name "www"
            ttl 3600
            cname_record "raymens.github.com"
        }
        dnsRecord {
            name "www"
            ttl 3600
            cname_record null
            target_resource (ResourceName "random")
        }
        dnsRecord {
            ttl 3600
            a_record "192.168.0.1"
        }
        dnsRecord {
            ttl 3600
            aaaa_record "1:1:1:1"
        }
    ]
}

let deployment = arm {
    location Location.NorthEurope

    add_resource dns

    //TODO: Assign resources here using the add_resource keyword
}
```

Now generates the following output:

```json
{
  "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
  "contentVersion": "1.0.0.0",
  "outputs": {},
  "parameters": {},
  "resources": [
    {
      "apiVersion": "2018-05-01",
      "location": "global",
      "name": "server",
      "properties": {
        "zoneType": "Public"
      },
      "type": "Microsoft.Network/dnszones"
    },
    {
      "apiVersion": "2018-05-01",
      "dependsOn": [
        "server"
      ],
      "name": "server/www",
      "properties": {
        "AAAARecord": {},
        "ARecord": {},
        "CNAMERecord": {
          "cname": "raymens.github.com"
        },
        "NsRecord": {},
        "PtrRecord": {},
        "TTL": 3600,
        "targetResource": {}
      },
      "type": "Microsoft.Network/dnszones/CNAME"
    },
    {
      "apiVersion": "2018-05-01",
      "dependsOn": [
        "server"
      ],
      "name": "server/@",
      "properties": {
        "AAAARecord": {},
        "ARecord": {
          "ipv4Address": "192.168.0.1"
        },
        "CNAMERecord": {},
        "NsRecord": {},
        "PtrRecord": {},
        "TTL": 3600,
        "targetResource": {}
      },
      "type": "Microsoft.Network/dnszones/A"
    }
  ]
}
```

raymens commented:
So I made some updates and created computation expressions for the different record types as the properties change.
It now also supports Record Sets and several fixes so that the ARM generated is correct.

```fsharp
let dns = dnsZone {
    name "raymens.com"
    add_records [
        cnameRecord {
            name "www2"
            ttl 3600
            cname "raymens.github.com"
        }
        aRecord {
            ttl 7200
            add_ipv4_addresses [ "192.168.0.1"; "192.168.0.2" ]
        }
        aaaaRecord {
            ttl 7200
            add_ipv6_addresses [ "100:100:100:100" ]
        }
    ]
}
```

Now generates:

```{
  "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
  "contentVersion": "1.0.0.0",
  "outputs": {},
  "parameters": {},
  "resources": [
    {
      "apiVersion": "2018-05-01",
      "location": "global",
      "name": "raymens.com",
      "properties": {
        "zoneType": "Public"
      },
      "type": "Microsoft.Network/dnszones"
    },
    {
      "apiVersion": "2018-05-01",
      "dependsOn": [
        "raymens.com"
      ],
      "name": "raymens.com/www2",
      "properties": {
        "AAAARecords": [],
        "ARecords": [],
        "CNAMERecords": {
          "cname": "raymens.github.com"
        },
        "NSRecords": [],
        "PTRRecords": [],
        "TTL": 3600,
        "targetResource": {}
      },
      "type": "Microsoft.Network/dnszones/CNAME"
    },
    {
      "apiVersion": "2018-05-01",
      "dependsOn": [
        "raymens.com"
      ],
      "name": "raymens.com/@",
      "properties": {
        "AAAARecords": [],
        "ARecords": [
          {
            "ipv4Address": "192.168.0.1"
          },
          {
            "ipv4Address": "192.168.0.2"
          }
        ],
        "CNAMERecords": {},
        "NSRecords": [],
        "PTRRecords": [],
        "TTL": 7200,
        "targetResource": {}
      },
      "type": "Microsoft.Network/dnszones/A"
    }
  ]
}
```

raymens commented:
I'm hitting an issue that causes records to be filtered out due to the distinct check on name and .NET type.
See:
```fsharp

    static member private AddResources(state:ArmConfig, resources:IArmResource list) =
        { state with
            Resources =
                state.Resources
                @ resources
                |> List.distinctBy(fun r -> r.ResourceName, r.GetType().Name) }
```

As I'm sharing the same resource name in the case of `@` records and the same underlying type (`DnsRecord`) is used not all resources go through this check.

There are 2 solutions I can think of right now:
1. Change my code to use different types instead of a shared type
2. Update the check to distinct by the ARM resource type instead of .NET type. Unsure if that's even possible.

raymens commented:
So I went with option 1 and am feeling quite confident about the generation now. Removing the draft property now :)

isaacabraham commented:
Cool, this looks great :-) @ninjarobot if you want to have a look at this and give feedback that'd be great, too.

isaacabraham commented:
@raymens the only other thing I can think of as opposed to builders for each "type" would be to use static methods on a class. This would allow you to request mandatory arguments and some as optional. But it's generally only good for "simple" types that have just a few fields.

raymens commented:
Well in the end I quite like the outcome with the different builders, even though it feels like there's quite some duplicate code right now. But for the end-user this feels like the best solution, putting in some extra effort is okay in that case.

Let me know if you have any comments and or questions, so I can resolve them quickly.

raymens commented:
So I was there was still some missing support for "name" of several DNS Records. I have added support but have defaulted to the apex domain (`@`). I can change that if you don't agree, might be better to be verbose and require setting it. However the Azure Portal also defaults to `@` in the case of an empty string.

I have also added some documentation, which might help you to comment as well.

isaacabraham commented:
@raymens I've merged from latest master into the PR. Could you add some release notes and also see if any tests are warranted?

isaacabraham commented on src/Farmer/Arm/Dns.fs:
```diff
@@ -0,0 +1,155 @@
+                {| this.Type.ResourceType.Create(this.Zone + this.Name, dependsOn = [ this.Zone ]) with
+                    properties =
+                        {| TTL = this.TTL
+                           targetResource = {| id = null |}
```
This field won't be emitted - JSON .NET is set up to remove any null values.

raymens replied:
Removed

isaacabraham commented on src/Farmer/Common.fs:
```diff
@@ -987,3 +987,7 @@ module Roles =
+
+module Dns =
+    type DnsRecordType = Unknown | A | AAAA | CName | NS | PTR | TXT | MX
+    // | Soa | Srv | CAA
```
Can we remove these commented out entries?

raymens replied:
Done

isaacabraham commented on src/Farmer/Arm/Dns.fs:
```diff
@@ -0,0 +1,155 @@
+        | PTR -> ptrRecord
+        | TXT -> txtRecord
+        | MX -> mxRecord
+        // | Soa -> soaRecord
```
Can we remove these commented out items?

raymens replied:
Done

raymens commented:
Thanks for the review.

I've added tests and fixed some things in the process.
Also made a deployment on Azure itself which succeeded.

isaacabraham commented on src/Farmer/Builders/Builders.Dns.fs:
```diff
@@ -0,0 +1,335 @@
+                else state.Name
+            TTL =
+                if state.TTL = None then failwith "You must set a TTL"
+                else state.TTL.Value
```
Don't use `.Value`. Instead use a pattern match.

isaacabraham replied:
Or `Option.defaultWith`.

raymens replied:
Went with the pattern match already :)

raymens replied:
Done

isaacabraham commented on src/Farmer/Builders/Builders.Dns.fs:
```diff
@@ -0,0 +1,335 @@
+    member __.Run(state : TxtRecordProperties) : DnsZoneRecordConfig =
+        { emptyRecord with
+            Name =
+                if state.Name = ResourceName.Empty then ResourceName "@"
```
Throw an exception here, rather than implicitly putting in a default value silently. Or, consider putting this into the default value within `Yield`.

raymens replied:
Will move it to `Yield`. Went with a silent default as it's what the Portal does as well and is common with these record types.

raymens replied:
Done

isaacabraham replied:
Great. If that's what the portal does, it's perhaps not so bad except that it's a "silent" change. I've tried to avoid that with other names as it can cause unintended surprises for authors. However this one seems somewhat more harmless.

raymens replied:
I'm fine with either cases, a default to "@" or it being required.

isaacabraham commented on src/Farmer/Arm/Dns.fs:
```diff
@@ -0,0 +1,151 @@
+                {| this.Type.ResourceType.Create(this.Zone + this.Name, dependsOn = [ this.Zone ]) with
+                    properties =
+                        {| TTL = this.TTL
+                           targetResource = {| id = this.TargetResource |}
```
Use `Option.toObj` to correctly serialize an `Option` value (`TargetResource`).

raymens replied:
That gives me the following error: `The type 'ResourceName' does not have 'null' as a proper value`. Not sure what to do.

isaacabraham replied:
Let me have a bash on the PR branch

isaacabraham replied:
Ok, so yes, that's right :-) That's because we should also be getting the ResourceName's `Value` out i.e. the string representation. This can then safely be optioned away:

```fsharp
targetResource = {| id = this.TargetResource |> Option.map(fun r -> r.Value) |> Option.toObj |}
```

**However**, this still leaves us with a potential object with nothing in it (I'm assuming that `targetResource` is indeed optional in the ARM template spec)! So if `targetResource` is None, you'll end up with simply

```json
"targetResource": {}
```

in the ARM template. Another way around this is to do the match at the level above e.g.

```fsharp
       targetResource =
        match this.TargetResource with
        | Some targetResource -> box {| id = targetResource.Value |}
        | None -> null
```

This will completely omit the `targetResource` field and object if the value in F# is None.

raymens replied:
According to the ARM template everything is optional hehe.

I've checked with our an existing DNS Zone in Azure to make sure, `"targetResource": {}` is valid when it's empty.

I guess the following would also work.
```fsharp
this.TargetResource
|> Option.map (fun x -> {| id = x.Value |}
|> Option.toObj
```

What solution shall we pick?

isaacabraham reviewed (changes requested):
Almost there :-) Just a few more changes required.

isaacabraham commented:
And it's in!
